README.md
# Affirm

**An Affirm API wrapper written in Elixir**

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed as:

  1. Add `affirm` to your list of dependencies in `mix.exs`:

  ```elixir
  def deps do
    [{:affirm, "~> 0.1.2"}]
  end
  ```

  2. Ensure `affirm` is started before your application:

  ```elixir
  def application do
    [applications: [:affirm]]
  end
  ```


CHANGELOG.md
0.1.2
------
#### Changes
* Bump HTTPoison dependency.


lib/test_helpers.ex
defmodule Affirm.TestHelpers do
  @doc "Creates a Bypass listener to act as a double for the actual Affirm API endpoints"
  @spec build_bypass :: any
  def build_bypass do
    bypass = Bypass.open()
    Application.put_env(:affirm, :url, "http://localhost:#{bypass.port}")
    bypass
  end

  @doc "Simulates that the provided Bypass listener is not accepting connections"
  @spec simulate_service_down(any) :: nil
  def simulate_service_down(bypass) do
    Bypass.down(bypass)
    nil
  end

  @spec simulate_service_response(any, Plug.Conn.status(), String.t(), (Plug.Conn.t() -> boolean)) :: no_return
  def simulate_service_response(bypass, status, body, fun) when is_function(fun) do
    Bypass.expect(bypass, fn conn ->
      if fun.(conn |> Plug.Conn.fetch_query_params()) do
        Plug.Conn.resp(conn, status, body)
      end
    end)
  end

  @spec successful_authorization_response_string(String.t()) :: String.t()
  def successful_authorization_response_string(amount, order_id \\ nil) do
    ~s(
        {"id":"ALO4-UVGR","created":"2016-03-18T19:19:04Z","currency":"USD","amount":#{amount},
        "auth_hold":6100,"payable":0,"void":false,"expires":"2016-04-18T19:19:04Z",
        "order_id":"#{order_id}","events":[{"created":"2014-03-20T14:00:33Z","currency":"USD",
        "id":"UI1ZOXSXQ44QUXQL","transaction_id":"TpR3Xrx8TkvuGio0","type":"auth"}],
        "details":{"items":{"sweater-a92123":{"sku":"sweater-a92123","display_name":"Sweater","qty":1,
        "item_type":"physical","item_image_url":"http://placehold.it/350x150","item_url":"http://placehold.it/350x150",
        "unit_price":5000}},"order_id":"#{order_id}","shipping_amount":400,"tax_amount":700,
        "shipping":{"name":{"full":"JohnDoe"},"address":{"line1":"325PacificAve","city":"SanFrancisco",
        "state":"CA","zipcode":"94112","country":"USA"}},"discounts":{"RETURN5":{"discount_amount":500,
        "discount_display_name":"Returningcustomer5%discount"},"PRESDAY10":{"discount_amount":1000,
        "discount_display_name":"President'sDay10%off"}}}}
    )
  end

  @spec successful_capture_response_string(String.t(), String.t()) :: String.t()
  def successful_capture_response_string(amount, fee) do
    ~s({"fee":#{fee},"created":"2016-03-18T00:03:44Z","order_id":"JKLM4321","currency":"USD",
    "amount":#{amount},"type":"capture","id":"O5DZHKL942503649","transaction_id":"6dH0LrrgUaMD7Llc"})
  end

  @spec successful_void_response_string() :: String.t()
  def successful_void_response_string() do
    ~s({"type":"void","id":"N5E9OXSIDJ8TKZZ9","transaction_id":"G9TqohxBlRPWTGB2",
    "created":"2014-03-17T22:52:16Z","order_id":"JLKM4321"})
  end

  @spec successful_refund_response_string(String.t(), String.t()) :: String.t()
  def successful_refund_response_string(amount, fee_refunded) do
    ~s({"created":"2014-03-18T19:20:30Z","fee_refunded":#{fee_refunded},"amount":#{amount},"type":"refund",
    "id":"OWA49MWUCA29SBVQ","transaction_id":"r86zdkHONPcaiVJJ"})
  end

  @spec failed_response_string(String.t()) :: String.t()
  def failed_response_string(code) do
    ~s({"status_code": 200, "code": "#{code}", "transaction_id": "12864241"})
  end

  @spec successful_read_response_string() :: String.t()
  def successful_read_response_string() do
    ~s({"id":"ALO4-UVGR","status":"auth","created":"2016-03-18T19:19:04Z","currency":"USD","amount":6100,
    "auth_hold":6100,"payable":0,"void":false,"expires":"2016-04-18T19:19:04Z","order_id":"JKLM4321",
    "events":[{"created":"2014-03-20T14:00:33Z","currency":"USD","id":"UI1ZOXSXQ44QUXQL",
    "transaction_id":"TpR3Xrx8TkvuGio0","type":"auth"}],"details":{"items":{"sweater-a92123":
    {"sku":"sweater-a92123","display_name":"Sweater","qty":1,"item_type":"physical",
    "item_image_url":"http://placehold.it/350x150","item_url":"http://placehold.it/350x150","unit_price":5000}},
    "order_id":"JKLM4321","shipping_amount":400,"tax_amount":700,"shipping":{"name":{"full":"John Doe"},"address":
    {"line1":"633 Folsom St","line2":"Floor 7","city":"San Francisco","state":"CA",
    "zipcode":"94112","country":"USA"}},"discounts":{"RETURN5":{"discount_amount":500,
    "discount_display_name":"Returning customer 5% discount"},
    "PRESDAY10":{"discount_amount":1000,"discount_display_name":"President's Day 10% off"}}}})
  end
end


test/affirm_test.exs
defmodule AffirmTest do
  use ExUnit.Case
  doctest Affirm
end


test/affirm/api_test.exs
defmodule Affirm.APITest do
  use ExUnit.Case
  import Affirm.TestHelpers

  alias Affirm.API

  setup context do
    {:ok,
      url: Application.put_env(:affirm, :url, "https://sandbox.affirm.com/api/v2/charges"),
      customer_id: Application.put_env(:affirm, :public_key, "bogus"),
      code: Application.put_env(:affirm, :private_key, "bogus"),
      affirm_bypass: (if context[:skip_bypass], do: nil, else: build_bypass)
    }
  end

  test """
  raises ConfigError when appropriate config vars are not set
  """ do
    Application.put_env(:affirm, :url, nil)

    assert_raise Affirm.ConfigError, "missing config for :url", fn ->
      API.authorize(%{})
    end
  end

  test """
  when the service endpoint is down
  returns an error
  """, %{affirm_bypass: affirm_bypass} do
    affirm_bypass
    |> simulate_service_down

    {:error, error} = API.authorize(%{})

    assert error == ":econnrefused"
  end

  test """
  authorize/1
  successfully POSTS and receives a response
  """, %{affirm_bypass: affirm_bypass} do
    response = successful_authorization_response_string("4500")

    params = %{
      checkout_token: "bogus",
      order_id: 1
    }

    affirm_bypass
    |> simulate_service_response(:ok, response, fn(conn) -> conn.method == "POST" end)

    {:ok, response} = API.authorize(params)

    assert response.amount == 4500
    assert response.type == "auth"
    refute is_nil(response.charge_id)
    refute is_nil(response.transaction_id)
  end

  test """
  authorize/1
  successfully POSTS and receives an error response
  when authorization is declined
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("auth-declined")

    params = %{
      checkout_token: "bogus",
      order_id: 1
    }

    affirm_bypass
    |> simulate_service_response(:ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.authorize(params)
    assert message == "Charge authorization hold declined."
  end

  test """
  capture/2
  successfully POSTS and receives a success response
  when given valid capture_id
  """, %{affirm_bypass: affirm_bypass} do
    response = successful_capture_response_string("4500", "450")

    # These params are optional
    params = %{
      order_id: "bogus",
      shipping_carrier: "USPS",
      shipping_confirmation: "1Z23223"
    }

    affirm_bypass
    |> simulate_service_response(:ok, response, fn(conn) -> conn.method == "POST" end)

    {:ok, response} = API.capture("CAPTURE_ID", params)
    assert response.amount == 4500
    assert response.fee == 450
    assert response.type == "capture"
    refute is_nil(response.id)
    refute is_nil(response.transaction_id)
    refute is_nil(response.created)
    refute is_nil(response.order_id)
  end

  test """
  capture/2
  successfully POSTS and receives a failure response
  when capture amount is greater than the auth
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("capture-greater-instrument")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.capture("BOGUS", %{})
    assert message == "Charges on this instrument cannot be captured for more than the authorization hold amount."
  end

  test """
  capture/2
  successfully POSTS and receives a failure response
  when capture amount does not equal than the auth
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("capture-unequal-instrument")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.capture("BOGUS", %{})
    assert message =~ "Charges on this instrument cannot be captured for an amount unequal to authorization"
  end

  test """
  capture/2
  successfully POSTS and receives a failure response
  when capture is already voided
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("capture-voided")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.capture("BOGUS", %{})
    assert message =~ "Cannot capture voided charge."
  end

  test """
  capture/2
  successfully POSTS and receives a failure response
  when capture is declined
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("capture-declined")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.capture("BOGUS", %{})
    assert message =~ "Charge capture declined."
  end

  test """
  capture/2
  successfully POSTS and receives a failure response
  when the capture limit is exceeded
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("capture-limit-exceeded")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.capture("BOGUS", %{})
    assert message =~ "Max capture amount on charge exceeded."
  end

  test """
  capture/2
  successfully POSTS and receives a failure response
  when the authorization is expired
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("expired-authorization")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.capture("BOGUS", %{})
    assert message == "Cannot capture expired charge authorization hold."
  end

  test """
  void/1
  successfully POSTS and receives a success response
  when given valid capture_id
  """, %{affirm_bypass: affirm_bypass} do
    response = successful_void_response_string()

    affirm_bypass
    |> simulate_service_response(:ok, response, fn(conn) -> conn.method == "POST" end)

    {:ok, response} = API.void("CAPTURE_ID")
    assert response.type == "void"
    refute is_nil(response.id)
    refute is_nil(response.transaction_id)
    refute is_nil(response.created)
    refute is_nil(response.order_id)
  end

  test """
  refund/2
  successfully POSTS and receives a success response
  when given valid capture_id and params
  """, %{affirm_bypass: affirm_bypass} do
    response = successful_refund_response_string("5000", "500")
    params = %{amount: "5000"}

    affirm_bypass
    |> simulate_service_response(:ok, response, fn(conn) -> conn.method == "POST" end)

    {:ok, response} = API.refund("CAPTURE_ID", params)
    assert response.type == "refund"
    assert response.amount == 5000
    assert response.fee_refunded == 500
    refute is_nil(response.id)
    refute is_nil(response.transaction_id)
    refute is_nil(response.created)
  end

  test """
  refund/2
  successfully POSTS and receives a failure response
  when the refund limit is exceeded
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("refund-exceeded")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.refund("BOGUS", %{})
    assert message == "Max refund amount exceeded."
  end

  test """
  refund/2
  successfully POSTS and receives a failure response
  when the auth has not been captured yet
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("refund-uncaptured")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.refund("BOGUS", %{})
    assert message == "Cannot refund a charge that has not been captured."
  end

  test """
  refund/2
  successfully POSTS and receives a failure response
  when the capture_id has been voided
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("refund-voided")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.refund("BOGUS", %{})
    assert message == "Cannot refund voided charge."
  end

  test """
  refund/2
  successfully POSTS and receives a failure response
  when the capture is more than 120 days old
  """, %{affirm_bypass: affirm_bypass} do
    response = failed_response_string("refund-expired")
    simulate_service_response(affirm_bypass, :ok, response, fn(conn) -> conn.method == "POST" end)

    {:error, %{message: message}} = API.refund("BOGUS", %{})
    assert message == "Charges on this instrument must be refunded within 120 days of capture."
  end

  test """
  read/2
  successfully GETs and receives a success response containing status
  when given valid capture_id
  """, %{affirm_bypass: affirm_bypass} do
    response = successful_read_response_string()

    affirm_bypass
    |> simulate_service_response(:ok, response, fn(conn) -> conn.method == "GET" end)

    {:ok, response} = API.read("CAPTURE_ID")
    assert response.status == "auth"
    refute Enum.empty?(response.events)
  end
end


mix.exs
defmodule Affirm.Mixfile do
  use Mix.Project

  @version "0.1.2"

  def project do
    [
      app: :affirm,
      version: @version,
      elixir: "~> 1.3",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: ExCoveralls],
      description: description(),
      package: package(),
      name: "Affirm",
      deps: deps()
    ]
  end

  def application do
    [applications: [:logger, :httpoison, :jason]]
  end

  defp description do
    """
    Affirm API library for Elixir.
    """
  end

  defp package do
    [
      maintainers: ["RevZilla", "Tyler Cain", "Steve DeGele"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/revzilla/affirm"},
      files: ~w(lib mix.exs README.md)
    ]
  end

  defp deps do
    [
      {:httpoison, "~> 1.1.1"},
      {:jason, "~> 1.0"},
      {:bypass, "~> 0.8", only: :test},
      {:excoveralls, "~> 0.7.4", only: :test},
      {:credo, "~> 0.8", only: [:dev, :test], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev}
    ]
  end
end


lib/affirm.ex
defmodule Affirm do
  @moduledoc """
  Affirm client library for Elixir.

  For api reference, please visit:
  https://docs.affirm.com/Integrate_Affirm/Direct_API

  Defines a `__using__` macro for ease of use inside implementation modules
  For example:
      defmodule AffirmService do
        use Affirm

        def attempt_capture(charge_id) do
          case capture(charge_id) do
            {:ok, response} -> # do success things
            {:error, message} -> # do failure things
          end
        end
      end
  """
  defmodule ConfigError do
    defexception [:message]

    @spec exception(String.t()) :: struct
    def exception(value) do
      message = "missing config for :#{value}"

      %ConfigError{message: message}
    end
  end

  defmacro __using__(_) do
    quote do
      import Affirm.API
    end
  end

  @doc """
  Convenience function for retrieving Affirm specfic environment values, but
  will raise an exception if values are missing.
  ## Example
      iex> Affirm.get_env(:random_value)
      ** (Affirm.ConfigError) missing config for :random_value
  """
  @spec get_env(atom) :: any
  def get_env(key) do
    Application.get_env(:affirm, key) || raise ConfigError, key
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :affirm, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:affirm, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/affirm/api.ex
defmodule Affirm.API do
  @moduledoc """
  Affirm provides several API endpoints... this module defines functions to
  interact with the transactional endpoints authorize, capture, void and refund.

  For api reference, please visit:
  https://docs.affirm.com/Integrate_Affirm/Direct_API
  """

  import Affirm.HTTP
  alias Affirm.Response

  @doc """
  Affirm's Authorize endpoint requires post data containing a `checkout_token`
  and takes an optional `order_id`. If the token is valid, the request will return
  a `charge_id` that should be saved for future transactions.
  """
  @spec authorize(map) :: {:ok, Response.t()} | {:error, String.t()}
  def authorize(params) do
    :post
    |> request("", params, headers, options)
    |> parse_response
  end

  @doc """
  Capture takes a `charge_id` and optional map of post data. Will return a `Affirm.Response`
  struct containing details of the successful capture transaction.
  """
  @spec capture(String.t(), map) :: {:ok, Response.t()} | {:error, String.t()}
  def capture(charge_id, params) do
    :post
    |> request("/#{charge_id}/capture", params, headers, options)
    |> parse_response
  end

  @doc """
  Capture takes a `charge_id`. Will return a `Affirm.Response`
  struct containing details of the successful void transaction.
  """
  @spec void(String.t()) :: {:ok, Response.t()} | {:error, String.t()}
  def void(charge_id) do
    :post
    |> request("/#{charge_id}/void", %{}, headers, options)
    |> parse_response
  end

  @doc """
  Refund takes a `charge_id` and a map containing a refund `amount`.
  Will return a `Affirm.Response` struct containing details of the successful return transaction.
  """
  @spec refund(String.t(), map) :: {:ok, Response.t()} | {:error, String.t()}
  def refund(charge_id, params) do
    :post
    |> request("/#{charge_id}/refund", params, headers, options)
    |> parse_response
  end

  @doc """
  Refund takes an optional `charge_id`.
  Will return an `Affirm.Response` struct containing current charge status.
  Used to prevent duplicate charge attempts.
  """
  @spec read(String.t()) :: {:ok, Response.t()} | {:error, String.t()}
  def read(charge_id) do
    :get
    |> request("/#{charge_id}", %{}, headers, options)
    |> parse_response(false)
  end

  @spec parse_response(tuple, boolean) :: {:ok, Response.t()} | {:error, String.t()}
  defp parse_response(response, is_transactional \\ true)

  defp parse_response({:ok, body}, is_transactional) do
    {:ok, Response.new(body, is_transactional)}
  end

  defp parse_response({:error, reason}, _), do: {:error, reason}

  @spec headers() :: list(tuple)
  defp headers do
    [
      {"Content-Type", "application/json"},
      {"Authorization", basic_auth(public_key, private_key)}
    ]
  end

  @spec options() :: keyword()
  defp options do
    [
      {:timeout, 15_000},
      {:recv_timeout, 15_000}
    ]
  end

  @spec basic_auth(binary, binary) :: binary
  defp basic_auth(user, pass) do
    "Basic " <> :base64.encode("#{user}:#{pass}")
  end

  defp public_key, do: Affirm.get_env(:public_key)
  defp private_key, do: Affirm.get_env(:private_key)
end


lib/affirm/response.ex
defmodule Affirm.Response do
  @moduledoc """
  Affirm defines all of the useful response data points in their API. This module
  houses a struct of those and provides a function for building the struct
  given a response map.
  """

  @type t :: %__MODULE__{
          amount: pos_integer,
          charge_id: String.t(),
          created: String.t(),
          currency: String.t(),
          fee: String.t(),
          fee_refunded: String.t(),
          id: String.t(),
          order_id: String.t(),
          transaction_id: String.t(),
          type: String.t(),
          message: String.t(),
          status: String.t(),
          events: list(map)
        }

  defstruct amount: nil,
            charge_id: nil,
            created: nil,
            currency: nil,
            fee: nil,
            fee_refunded: nil,
            id: nil,
            order_id: nil,
            transaction_id: nil,
            type: nil,
            message: nil,
            status: nil,
            events: nil

  @doc """
  Function builds a response object based on parsed json return map
  send back from Affirm.
  """
  @spec new(map, boolean) :: t
  def new(response, is_transactional \\ true)

  def new(%{events: events, amount: amount, id: charge_id, order_id: order_id}, true) do
    events
    |> hd
    |> new
    |> dump_into_struct(%{amount: amount, charge_id: charge_id, order_id: order_id})
  end

  def new(nil, _), do: %__MODULE__{}

  def new(response, _) do
    case Enum.empty?(response) do
      false -> %__MODULE__{} |> dump_into_struct(response)
      _ -> %__MODULE__{}
    end
  end

  @spec dump_into_struct(struct, map) :: t
  defp dump_into_struct(mod, map_response) do
    Enum.reduce(Map.keys(map_response), mod, fn atom, acc ->
      value = Map.get(map_response, atom)

      verify_atom =
        case is_binary(atom) do
          true -> String.to_atom(atom)
          _ -> atom
        end

      Map.put(acc, verify_atom, value)
    end)
  end
end


lib/http.ex
defmodule Affirm.HTTP do
  @moduledoc """
  Base client for all server interaction, used by all endpoint specific
  modules. This request wrapper coordinates the remote server, headers,
  authorization and SSL options.

  This uses `HTTPoison.Base`, so all of the typical HTTP verbs are availble.

  Sample Affirm config:
    config :affirm,
      url: "base_api_url",
      public_key: "affirm_supplied_public_key",
      private_key: "affirm_supplied_private_key"
  """
  use HTTPoison.Base
  alias HTTPoison.Response
  alias HTTPoison.AsyncResponse

  @spec request(atom, binary, binary, headers, Keyword.t()) ::
          {:ok, Response.t() | AsyncResponse.t()} | {:error, integer, Response.t()} | {:error, atom | String.t()}
  def request(method, url, body, headers \\ [], options \\ []) do
    method
    |> super(full_url(url), body, headers, options)
    |> process_response
  end

  @spec full_url(binary) :: binary
  def full_url(url) do
    Affirm.get_env(:url) <> url
  end

  ## HTTPoison Callbacks

  @doc false
  def process_request_body(body) when body == "" or body == %{}, do: ""

  def process_request_body(body) do
    Jason.encode!(body)
  end

  @doc false
  def process_response_body(body) do
    case Jason.decode(body) do
      {:ok, parsed_body} ->
        Enum.reduce(parsed_body, %{}, fn {key, val}, acc -> Map.put(acc, String.to_atom(key), val) end)

      {:error, _} ->
        body
    end
  end

  @doc false
  def process_response({:ok, %{status_code: 200, body: %{code: error_code} = body}}) do
    case fetch_error_code(error_code) do
      nil -> {:error, Affirm.Response.new(%{message: body["message"]})}
      message -> {:error, Affirm.Response.new(%{message: message})}
    end
  end

  def process_response({:ok, %{status_code: code, body: body}})
      when code >= 200 and code <= 399,
      do: {:ok, body}

  def process_response({:ok, %{status_code: 400}}), do: {:error, :invalid_request}
  def process_response({:ok, %{status_code: 401}}), do: {:error, :unauthorized}
  def process_response({:ok, %{status_code: 404}}), do: {:error, :not_found}
  def process_response({:ok, %{body: body}}), do: {:error, body}
  def process_response({:error, ":econnrefused"}), do: {:error, :econnrefused}
  def process_response({_code, %HTTPoison.Error{reason: reason}}), do: {:error, inspect(reason)}

  @doc """
  Provides a mapping from returned error code to error String.
  Affirm also provides this message in the returned response, but this
  provides an easier way of mocking/testing.
  """
  @spec fetch_error_code(String.t()) :: String.t()
  def fetch_error_code(error_code) do
    # credo:disable-for-lines:2 Credo.Check.Readability.MaxLineLength
    code_map = %{
      "auth-declined" => "Charge authorization hold declined.",
      "capture-greater-instrument" =>
        "Charges on this instrument cannot be captured for more than the authorization hold amount.",
      "capture-unequal-instrument" =>
        "Charges on this instrument cannot be captured for an amount unequal to authorization hold amount.",
      "capture-voided" => "Cannot capture voided charge.",
      "partial-capture-instrument" => "Charges on this instrument cannot be partially captured.",
      "refund-exceeded" => "Max refund amount exceeded.",
      "refund-uncaptured" => "Cannot refund a charge that has not been captured.",
      "refund-voided" => "Cannot refund voided charge.",
      "capture-declined" => "Charge capture declined.",
      "capture-limit-exceeded" => "Max capture amount on charge exceeded.",
      "expired-authorization" => "Cannot capture expired charge authorization hold.",
      "refund-expired" => "Charges on this instrument must be refunded within 120 days of capture.",
      "financial-product-invalid" => "Please provide a valid financial product key.",
      "invalid_field" => "An input field resulted in invalid request."
    }

    code_map[error_code]
  end
end
